src/main/kotlin/org/rust/ide/intentions/ImportNameIntention.kt

```diff
@@ -0,0 +1,232 @@
+import com.intellij.codeInsight.navigation.NavigationUtil
+import com.intellij.ide.util.DefaultPsiElementCellRenderer
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.popup.PopupStep
+import com.intellij.openapi.ui.popup.util.BaseListPopupStep
+import com.intellij.psi.PsiElement
+import com.intellij.ui.popup.list.ListPopupImpl
+import com.intellij.ui.popup.list.PopupListElementRenderer
+import com.intellij.util.ObjectUtils
+import com.intellij.util.ui.UIUtil
+import org.rust.cargo.icons.CargoIcons
+import org.rust.cargo.project.workspace.PackageOrigin
+import org.rust.cargo.util.AutoInjectedCrates
+import org.rust.ide.icons.RsIcons
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.stubs.index.RsNamedElementIndex
+import org.rust.lang.core.types.ty.TyPrimitive
+import org.rust.openapiext.Testmark
+import java.awt.BorderLayout
+import java.awt.Component
+import javax.swing.*
+
+class ImportNameIntention : RsElementBaseIntentionAction<ImportNameIntention.Context>(), HighPriorityAction {
+
+    override fun getText() = "Import"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val path = element.ancestorStrict<RsPath>() ?: return null
+        if (TyPrimitive.fromPath(path) != null || path.reference.resolve() != null) return null
+
+        val parent = path.path
+        val parentRes = parent?.reference?.resolve()
+        if (parent == null || parentRes is RsMod || parentRes is RsEnumItem) {
+            val basePath = getBasePath(path)
+            val pathMod = path.containingMod
+            val pathSuperMods = HashSet(pathMod.superMods)
+            // TODO: support reexports
+            val candidates = RsNamedElementIndex.findElementsByName(project, basePath.referenceName)
+                .filterIsInstance<RsQualifiedNamedElement>()
+                .filter { (it as? RsVisible)?.isVisibleIn(pathSuperMods) ?: false }
+            if (candidates.isNotEmpty()) {
+                return Context(path, candidates)
+            }
+        }
+
+        return null
+    }
+
+    // Semantic signature of method is `RsVisible.isVisibleIn(mod: RsMod)`
+    // but in our case `mod` is always same and `mod` needs only to get set of its super mods
+    // so we pass `superMods` instead of `mod` for optimization
+    private fun RsVisible.isVisibleIn(superMods: Set<RsMod>): Boolean {
+        val parentMod = (if (this is RsMod) `super` else containingMod) ?: return false
+        val ourSuperMods = parentMod.superMods
+
+        // try to find latest common ancestor module of `parentMod` and `mod` in module tree
+        // we need to do it because we can use direct child items of any super mod with any visibility
+        val lca = ourSuperMods.find { it in superMods }
```

Comment:
Interesting! I suppose that we might need an `lca` when we are figuring out the optimal import path. So perhaps this function should return not only a boolean result, but also a path (perhaps as a string) which should be used in the corresponding `use` item?

Reply:
Ah, so we need to handle extern crates as well... Perhaps `RsVisible.canBeImported() -> ImportDescription?` is needed here, where Import Description is a sealed class?